# Importing Libraries

In [4]:
import pandas as pd
import numpy as np
import os

In [5]:
# create a system path
path = r'/Users/priyankamittal/Desktop/Data Analytics/CareerFoundry/Immersion/Python/07-2023 Instacart Basket Analysis Copy'

In [6]:
#read the pickle file
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [7]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0   2539329        1    prior             1                  2   
1   2398795        1    prior             2                  3   
2    473747        1    prior             3                  3   
3   2254736        1    prior             4                  4   
4    431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

      price_range_loc     busiest_day         busiest_days  \
0  Mid-range Products  Regularly busy  Regularly busy days   
1  Mid-range Products  Regularly busy      Least Busy days   
2  Mid-range Products  Regularly busy      Least Busy days   
3  Mid-range Products      Least Busy      Least Busy days   
4  Mid-range Products      Least Busy      Least Busy days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

In [8]:
# Always check the number of rows and columns
ords_prods_merge.shape

(32434212, 19)

In [9]:
# creating a subset
df = ords_prods_merge[:1000000]

In [10]:
df.shape

(1000000, 19)

In [11]:
# Grouping product name but out put will not show as desried.for aggregate () is required 
df.groupby('product_name')

In [12]:
# Create a group by and use aggregate ()
df.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [13]:
# There are some aggregations that can be conducted without use of the agg() function. 
#For instance, the command above could be replaced with a command that uses the mean() function to 
#achieve the same results:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [14]:
#DOT NOTATION VS. SQUARE BRACKETS
# Rather than including the column name within square brackets, as in the example above, you could also set the column name off with a dot, for instance:
df.groupby('department_id').order_number.mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [15]:
# Using multiple agg function
df.groupby('department_id').agg({'order_number' : ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Create a loyality flag

In [16]:
# Now, let’s map this task onto the three-step process introduced earlier:

# 1. Split the data into groups based on the “user_id” column.
# 2. Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
# 3. Create a new column, “max_order,” into which you’ll place the results of your aggregation.

In [17]:
# Flag means meeting certain criteria.'np' meand numply
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

# df_orders_products_merged['max_order'] = df_orders_products_merged.groupby(['user_id'])['order_number'].transform(np.max)

# Let’s run through each piece of the code in turn:
# First, a new column called “max_order” is created, which will be what stores the maximum order number for each user (step 3).
# Then, the df_orders_products_merged dataframe is grouped by the “user_id” column (step 1).
# And finally, the transform() function is applied on the “order_number” column with the np.max argument (step 2).

In [18]:
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
5    3367565        1    prior             6                  2   
6     550135        1    prior             7                  1   
7    3108588        1    prior             8                  1   
8    2295261        1    prior             9                  1   
9    2550362        1    prior            10                  4   
10   2968173       15    prior            15                  1   
11   1870022       15    prior            17                  2   
12   1911383       15    prior            18                  2   
13   2715276       15    prior            21                  1   
14    487368       15    prior            22                  1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
5           1         Soda        77              7     9.0   both   
6           1         Soda        77              7     9.0   both   
7           1         Soda        77              7     9.0   both   
8           1         Soda        77              7     9.0   both   
9           1         Soda        77              7     9.0   both   
10          0         Soda        77              7     9.0   both   
11          1         Soda        77              7     9.0   both   
12          1         Soda        77              7     9.0   both   
13          1         Soda        77              7     9.0   both   
14          1         Soda        77              7     9.0   both   

       price_range_loc     busiest_day         busiest_days  \
0   Mid-range Products  Regularly busy  Regularly busy days   
1   Mid-range Products  Regularly busy      Least Busy days   
2   Mid-range Products  Regularly busy      Least Busy days   
3   Mid-range Products      Least Busy      Least Busy days   
4   Mid-range Products      Least Busy      Least Busy days   
5   Mid-range Products  Regularly busy  Regularly busy days   
6   Mid-range Products  Regularly busy         Busiest days   
7   Mid-range Products  R

In [19]:
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
..       ...      ...      ...           ...                ...   
95   3226575      360    prior             1                  5   
96   1469869      377    prior             3                  5   
97   1927023      387    prior             2                  4   
98    858092      420    prior             4                  1   
99    119314      420    prior             5                  3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
..        ...          ...       ...            ...     ...    ...   
95          0         Soda        77              7     9.0   both   
96          0         Soda        77              7     9.0   both   
97          0         Soda        77              7     9.0   both   
98          0         Soda        77              7     9.0   both   
99          1         Soda        77              7     9.0   both   

       price_range_loc     busiest_day         busiest_days  \
0   Mid-range Products  Regularly busy  Regularly busy days   
1   Mid-range Products  Regularly busy      Least Busy days   
2   Mid-range Products  Regularly busy      Least Busy days   
3   Mid-range Products      Least Busy      Least Busy days   
4   Mid-range Products      Least Busy      Least Busy days   
..                 ...             ...                  ...   
95  Mid-range Products  Regularly busy  Regularly busy days   
96  Mid-range Products  Regularly busy  Regularly busy days   
97  Mid-range Products      Least Busy      Least Busy days   
98  Mid-range Products  Regularly busy         Busiest days   
99  Mid-range Products  Regularly busy      Least Busy days   

   busiest_period_of_day  max_order  
0         Average orders         10  
1         Average orders         10  
2            Most orders         10  
3         Average orders         10  
4            Most orders         10  
..                   ...        ...  
95           Most orders          3  
96        Average orders          3  
97           Most orders          8  
98        Average orders         22  
99        Average orders         22  

[100 rows x 20 columns]

In [20]:
pd.options.display.max_rows = None

In [21]:
ords_prods_merge.head(100)

order_id  user_id eval_set  order_number  order_day_of_week  \
0    2539329        1    prior             1                  2   
1    2398795        1    prior             2                  3   
2     473747        1    prior             3                  3   
3    2254736        1    prior             4                  4   
4     431534        1    prior             5                  4   
5    3367565        1    prior             6                  2   
6     550135        1    prior             7                  1   
7    3108588        1    prior             8                  1   
8    2295261        1    prior             9                  1   
9    2550362        1    prior            10                  4   
10   2968173       15    prior            15                  1   
11   1870022       15    prior            17                  2   
12   1911383       15    prior            18                  2   
13   2715276       15    prior            21                  1   
14    487368       15    prior            22                  1   
15   2293453       19    prior             2                  5   
16   1973799       19    prior             5                  6   
17    532817       19    prior             7                  4   
18   1573906       21    prior            10                  3   
19   1593000       31    prior            10                  3   
20   2231262       31    prior            17                  3   
21   2580647       43    prior             6                  4   
22   2187180       43    prior             9                  4   
23   2497897       52    prior             1                  1   
24   1318871       52    prior             2                  1   
25   1261384       52    prior             3                  1   
26    580568       52    prior             5                  2   
27   2428073       52    prior             7                  2   
28    342306       52    prior             9                  1   
29   1520353       52    prior            10                  2   
30   2542086       52    prior            11                  2   
31    944694       52    prior            12                  1   
32   1268191       52    prior            15                  2   
33   1122089       52    prior            17                  2   
34   1498922       52    prior            18                  2   
35    180919       52    prior            19                  3   
36   2695875       52    prior            21                  5   
37   3205960       67    prior             1                  3   
38    970922       67    prior             2                  1   
39    596585       67    prior             3                  1   
40   2037087       67    prior             4                  2   
41   2474111       67    prior             5                  3   
42   1568481       67    prior             6                  3   
43   2048527       67    prior             8                  2   
44   2389031       67    prior             9                  2   
45   2378562       67    prior            10                  1   
46   1515698       67    prior            11                  1   
47   1928050       67    prior            12                  2   
48   1949274       67    prior            13                  1   
49    551869       67    prior            14                  1   
50    768572       67    prior            16                  4   
51    150528       67    prior            18                  5   
52     29758       67    prior            19                  0   
53   1558976       67    prior            21                  1   
54    863974       67    prior            22                  1   
55   2308399       67    prior            23                  1   
56   2759128       81    prior             3                  1   
57   1812985       81    prior             5                  2   
58   1889835       82    prior             1       

### Deriving Columns with loc()

In [22]:
# To create your flag, you’ll need some criteria. You can use the following:

# 1. If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
# 2. If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
# 3. If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [23]:
#Creating flags
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'Loyality Flag'] = 'Loyal Customer'

In [24]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'Loyality Flag'] = 'Regular Customer'

In [25]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'Loyality Flag'] = 'New Customer'

In [26]:
ords_prods_merge['Loyality Flag'].value_counts(dropna = False)

Regular Customer    15891077
Loyal Customer      10293737
New Customer         6249398
Name: Loyality Flag, dtype: int64

In [27]:
# You learned already that to access a single column in a dataframe, you can index it like so: df['column']

In [28]:
# # check the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' colu
ords_prods_merge[['user_id', 'Loyality Flag', 'order_number']].head(60)

# The first set of brackets is the same as the brackets used in df['column'] above,
# while the second set of brackets is indicating that what you’re indexing is a list of multiple columns (and lists are always included within brackets).
# This is why there are two sets of brackets.

user_id     Loyality Flag  order_number
0         1      New Customer             1
1         1      New Customer             2
2         1      New Customer             3
3         1      New Customer             4
4         1      New Customer             5
5         1      New Customer             6
6         1      New Customer             7
7         1      New Customer             8
8         1      New Customer             9
9         1      New Customer            10
10       15  Regular Customer            15
11       15  Regular Customer            17
12       15  Regular Customer            18
13       15  Regular Customer            21
14       15  Regular Customer            22
15       19      New Customer             2
16       19      New Customer             5
17       19      New Customer             7
18       21  Regular Customer            10
19       31  Regular Customer            10
20       31  Regular Customer            17
21       43  Regular Customer             6
22       43  Regular Customer             9
23       52  Regular Customer             1
24       52  Regular Customer             2
25       52  Regular Customer             3
26       52  Regular Customer             5
27       52  Regular Customer             7
28       52  Regular Customer             9
29       52  Regular Customer            10
30       52  Regular Customer            11
31       52  Regular Customer            12
32       52  Regular Customer            15
33       52  Regular Customer            17
34       52  Regular Customer            18
35       52  Regular Customer            19
36       52  Regular Customer            21
37       67  Regular Customer             1
38       67  Regular Customer             2
39       67  Regular Customer             3
40       67  Regular Customer             4
41       67  Regular Customer             5
42       67  Regular Customer             6
43       67  Regular Customer             8
44       67  Regular Customer             9
45       67  Regular Customer            10
46       67  Regular Customer            11
47       67  Regular Customer            12
48       67  Regular Customer            13
49       67  Regular Customer            14
50       67  Regular Customer            16
51       67  Regular Customer            18
52       67  Regular Customer            19
53       67  Regular Customer            21
54       67  Regular Customer            22
55       67  Regular Customer            23
56       81      New Customer             3
57       81      New Customer             5
58       82  Regular Customer             1
59       82  Regular Customer             3

In [29]:
# And that brings you to the end of the three-step process!
# 1. You started by grouping your data,
# 2. followed up by applying an aggregation function,
# 3. and finished everything off by using the resulting aggregation to create a flag column

### Task 4.8

### Step 1
#### Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [30]:
#done

### Step 2
####In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.¶

In [31]:
ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3

#### Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

In [32]:
df.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

#### Answer: The average result of order_number from the subset (1000000) is greater than the average result of entire dataframe (except the department_id: 17) 

### Step 4


In [33]:
#### Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [34]:
## check the output [31]

### Step 5

#### Determine the difference between the spending habits of the three types of customers you identified.
#### Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.`

In [35]:
print('The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:')
ords_prods_merge.groupby('prices').agg({'prices': ['mean', 'min', 'max','count']})

The mean, min and max of prices grouped by the loyalty_flag column in df_orders_products_merged:


prices                          
            mean      min      max   count
prices                                    
1.0          1.0      1.0      1.0  115143
1.1          1.1      1.1      1.1  242782
1.2          1.2      1.2      1.2  256276
1.3          1.3      1.3      1.3  425691
1.4          1.4      1.4      1.4  392209
1.5          1.5      1.5      1.5  154527
1.6          1.6      1.6      1.6  299352
1.7          1.7      1.7      1.7  213967
1.8          1.8      1.8      1.8  321184
1.9          1.9      1.9      1.9  156179
2.0          2.0      2.0      2.0  286259
2.1          2.1      2.1      2.1  206004
2.2          2.2      2.2      2.2  160243
2.3          2.3      2.3      2.3  242051
2.4          2.4      2.4      2.4  196042
2.5          2.5      2.5      2.5  246421
2.6          2.6      2.6      2.6  211936
2.7          2.7      2.7      2.7  199785
2.8          2.8      2.8      2.8  153979
2.9          2.9      2.9      2.9  264700
3.0          3.0      3.0      3.0  192650
3.1          3.1      3.1      3.1  249256
3.2          3.2      3.2      3.2  253782
3.3          3.3      3.3      3.3  238041
3.4          3.4      3.4      3.4  223571
3.5          3.5      3.5      3.5  222786
3.6          3.6      3.6      3.6  300127
3.7          3.7      3.7      3.7  222159
3.8          3.8      3.8      3.8  315426
3.9          3.9      3.9      3.9  212805
4.0          4.0      4.0      4.0  447272
4.1          4.1      4.1      4.1  268269
4.2          4.2      4.2      4.2  245755
4.3          4.3      4.3      4.3  162367
4.4          4.4      4.4      4.4  357026
4.5          4.5      4.5      4.5  367223
4.6          4.6      4.6      4.6  144069
4.7          4.7      4.7      4.7  221048
4.8          4.8      4.8      4.8  193898
4.9          4.9      4.9      4.9  318220
5.0          5.0      5.0      5.0  225904
5.1          5.1      5.1      5.1  224797
5.2          5.2      5.2      5.2  251509
5.3          5.3      5.3      5.3  284243
5.4          5.4      5.4      5.4  206769
5.5          5.5      5.5      5.5  250136
5.6          5.6      5.6      5.6  183638
5.7          5.7      5.7      5.7  323997
5.8          5.8      5.8      5.8  241047
5.9          5.9      5.9      5.9  208900
6.0          6.0      6.0      6.0  196985
6.1          6.1      6.1      6.1  244209
6.2          6.2      6.2      6.2  178656
6.3          6.3      6.3      6.3  438467
6.4          6.4      6.4      6.4  206468
6.5          6.5      6.5      6.5  278357
6.6          6.6      6.6      6.6  204936
6.7          6.7      6.7      6.7  379126
6.8          6.8      6.8      6.8  355246
6.9          6.9      6.9      6.9  314196
7.0          7.0      7.0      7.0  282884
7.1          7.1      7.1      7.1  140813
7.2          7.2      7.2      7.2  223120
7.3          7.3      7.3      7.3  411924
7.4          7.4      7.4      7.4  233506
7.5          7.5      7.5      7.5  239261
7.6          7.6      7.6      7.6  113878
7.7          7.7      7.7      7.7  190070
7.8          7.8      7.8      7.8  186923
7.9          7.9      7.9      7.9  342988
8.0          8.0      8.0      8.0  151542
8.1          8.1      8.1      8.1  176805
8.2          8.2      8.2      8.2  404015
8.3          8.3      8.3      8.3  276230
8.4          8.4      8.4      8.4  227493
8.5          8.5      8.5      8.5  139244
8.6          8.6      8.6      8.6  254009
8.7          8.7      8.7      8.7  166742
8.8          8.8      8.8      8.8  318091
8.9          8.9      8.9      8.9  219412
9.0          9.0      9.0      9.0  228576
9.1          9.1      9.1      9.1  320782
9.2          9.2      9.2      9.2  182444
9.3          9.3      9.3      9.3  168917
9.4          9.4      9.4      9.4  197179
9.5          9.5      9.5      9.5  165789
9.6          9.6      9.6      9.6  216626
9.7          9.7      9.7      9.7  120777
9.8          9.8      9.8      9.8  194754
9.9          9.9      9.9      9.9  169263
10.0       

In [36]:
ords_prods_merge.groupby('prices').agg({'prices': ['mean']})

prices
            mean
prices          
1.0          1.0
1.1          1.1
1.2          1.2
1.3          1.3
1.4          1.4
1.5          1.5
1.6          1.6
1.7          1.7
1.8          1.8
1.9          1.9
2.0          2.0
2.1          2.1
2.2          2.2
2.3          2.3
2.4          2.4
2.5          2.5
2.6          2.6
2.7          2.7
2.8          2.8
2.9          2.9
3.0          3.0
3.1          3.1
3.2          3.2
3.3          3.3
3.4          3.4
3.5          3.5
3.6          3.6
3.7          3.7
3.8          3.8
3.9          3.9
4.0          4.0
4.1          4.1
4.2          4.2
4.3          4.3
4.4          4.4
4.5          4.5
4.6          4.6
4.7          4.7
4.8          4.8
4.9          4.9
5.0          5.0
5.1          5.1
5.2          5.2
5.3          5.3
5.4          5.4
5.5          5.5
5.6          5.6
5.7          5.7
5.8          5.8
5.9          5.9
6.0          6.0
6.1          6.1
6.2          6.2
6.3          6.3
6.4          6.4
6.5          6.5
6.6          6.6
6.7          6.7
6.8          6.8
6.9          6.9
7.0          7.0
7.1          7.1
7.2          7.2
7.3          7.3
7.4          7.4
7.5          7.5
7.6          7.6
7.7          7.7
7.8          7.8
7.9          7.9
8.0          8.0
8.1          8.1
8.2          8.2
8.3          8.3
8.4          8.4
8.5          8.5
8.6          8.6
8.7          8.7
8.8          8.8
8.9          8.9
9.0          9.0
9.1          9.1
9.2          9.2
9.3          9.3
9.4          9.4
9.5          9.5
9.6          9.6
9.7          9.7
9.8          9.8
9.9          9.9
10.0        10.0
10.1        10.1
10.2        10.2
10.3        10.3
10.4        10.4
10.5        10.5
10.6        10.6
10.7        10.7
10.8        10.8
10.9        10.9
11.0        11.0
11.1        11.1
11.2        11.2
11.3        11.3
11.4        11.4
11.5        11.5
11.6        11.6
11.7        11.7
11.8        11.8
11.9        11.9
12.0        12.0
12.1        12.1
12.2        12.2
12.3        12.3
12.4        12.4
12.5        12.5
12.6        12.6
12.7        12.7
12.8        12.8
12.9        12.9
13.0        13.0
13.1        13.1
13.2        13.2
13.3        13.3
13.4        13.4
13.5        13.5
13.6        13.6
13.7        13.7
13.8        13.8
13.9        13.9
14.0        14.0
14.1        14.1
14.2        14.2
14.3        14.3
14.4        14.4
14.5        14.5
14.6        14.6
14.7        14.7
14.8        14.8
14.9        14.9
15.0        15.0
15.1        15.1
15.2        15.2
15.3        15.3
15.4        15.4
15.5        15.5
15.6        15.6
15.7        15.7
15.8        15.8
15.9        15.9
16.0        16.0
16.1        16.1
16.2        16.2
16.3        16.3
16.4        16.4
16.5        16.5
16.6        16.6
16.7        16.7
16.8        16.8
16.9        16.9
17.0        17.0
17.1        17.1
17.2        17.2
17.3        17.3
17.4        17.4
17.5        17.5
17.6        17.6
17.7        17.7
17.8        17.8
17.9        17.9
18.0        18.0
18.1        18.1
18.2        18.2
18.3        18.3
18.4        18.4
18.5        18.5
18.6        18.6
18.7        18.7
18.8        18.8
18.9        18.9
19.0        19.0
19.1        19.1
19.2        19.2
19.3        19.3
19.4        19.4
19.5        19.5
19.6        19.6
19.7        19.7
19.8        19.8
19.9        19.9
20.0        20.0
20.1        20.1
20.2        20.2
20.3        20.3
20.4        20.4
20.5        20.5
20.6        20.6
20.7        20.7
20.8        20.8
20.9        20.9
21.0        21.0
21.1        21.1
21.3        21.3
21.4        21.4
21.5        21.5
21.6        21.6
21.7        21.7
21.8        21.8
21.9        21.9
22.0        22.0
22.1        22.1
22.2        22.2
22.3        22.3
22.4        22.4
22.5        22.5
22.6        22.6
22.7        22.7
22.8        22.8
22.9        22.9
23.0        23.0
23.1        23.1
23.2        23.2
23.3        23.3
23.4        23.4
23.5        23.5
23.6        23.6
23.7        23.7
23.8        23.8
23.9        23.9
24.0        24.0
24.1        24.1
24.2        24.2
24.3        24.

In [37]:
ords_prods_merge.agg({'prices': ['mean']})

prices
mean  11.981144

In [38]:
#### Observation:
#### 1. It is interesting to see that new customers have a higher average purchase price than the other 2 customer categories and even loyal customers have a lower average purchase than the others
#### 2. The maximum price here seems odd compared to the mean value itself. The value of 1499.0 has skewed the distribution so further analysis for this value will be performed.

In [39]:
# Run descriptive statistics on the 'prices' column

print('Descriptive statistics of the prices column:')
ords_prods_merge['prices'].describe()

Descriptive statistics of the prices column:


count    3.243421e+07
mean     1.198114e+01
std      4.954311e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

#### Observation:
#### 1. The average (mean) price is 10.6. and the maximum price is 14.9 through this we can see that there is a huge difference in this data
#### 2. The standard deviation has been confirmed which has a std of  20.2 indicating that we have a lot of variation in this data (especially large values that are higher than the mean)

In [40]:
# Check for abnormal values in the 'prices' column

##ords_prods_merge['prices'].sort_values(ascending = False)

In [41]:
print('Check for abnormal values in the prices column:')

ords_prods_merge['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in the prices column:


29192575    99999.0
10030345    14900.0
18529046       25.0
25785942       24.9
28452491       24.8
14495433       24.7
23398589       24.6
19834898       24.5
21370901       24.4
23988774       24.3
7073293        24.2
13037681       24.1
15686466       24.0
10276407       23.9
19838750       23.8
16567274       23.7
10350371       23.6
25821133       23.5
24697204       23.4
18402120       23.3
5269126        23.2
10391784       23.1
9286103        23.0
21438101       22.9
25580959       22.8
31818826       22.7
23114973       22.6
29413807       22.5
20364464       22.4
7016532        22.3
12976869       22.2
15472566       22.1
25090431       22.0
24631846       21.9
20389607       21.8
19345153       21.7
15008750       21.6
29743151       21.5
30393304       21.4
19174773       21.3
7060301        21.1
32416885       21.0
24782324       20.9
28677673       20.8
12275275       20.7
12975696       20.6
24916798       20.5
13252406       20.4
25616516       20.3
31390895       20.2


#### Observation: The 14.900 prices look weird (that's a lot higher than the others)

In [42]:
# Get the output of df_orders_products_merged when the value in the 'prices' column is 99999 or 14900

max_price = ords_prods_merge.loc[ords_prods_merge['prices'].isin([14900])]

In [43]:
# ords_prods_merge["prices"].replace(14900.0, 14.90, inplace=True)

In [ ]:
ords_prods_merge.loc[ords_prods_merge['prices'].isin([14.90])]

In [ ]:
pd.options.display.max_columns = None

In [ ]:
# Sample output:

print('Sample output of ords_prods_merge when the value in the prices column is 14900:')
max_price.sample(5)

In [ ]:
# Descriptive statistics on max_price

print('Descriptive statistics on max_price:')
max_price.describe()

# I need to scroll to the right to find the price, department_id and product_id

In [ ]:
#### Observation: Incredibly high prices (14.900 ) is from the department_id 16 with product_id 21553

In [ ]:
pd.options.display.max_columns = None

In [ ]:
# Find the product_name with product_id 21553

ords_prods_merge.loc[ords_prods_merge['product_id'].isin([21553])]

In [ ]:
#### Observation: 
#### 1. The price of 14.900 comes from Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)

In [ ]:
pd.options.display.max_columns = None

In [ ]:
# Create a filter 'dairy' for records with the department_id value of 16, 
# and then 'dairy_without_weird' to exclude records (without very high prices)
# using ~ to exclude the data

dairy = ords_prods_merge.loc[ords_prods_merge['department_id'].isin([16])]
dairy_without_weird = dairy.loc[~dairy['prices'].isin([14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
dairy_without_weird.sample(5)

In [ ]:
# Check to see if the filter is working properly

print('Maximum price in the dairy_without_weird filter:')
dairy_without_weird['prices'].max()

In [ ]:
#### Observation: After excluding unusual values, the max price becomes more reasonable


In [ ]:
# Second: Observation on Lowfat 2% Milkfat Cottage Cheese (with product_id 21553)
# Create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird

cottage_cheese = dairy_without_weird.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [ ]:
print('Sample output from cottage_cheese:')
cottage_cheese.sample(10)

In [ ]:
# Quick statistics in cottage_cheese

print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

In [ ]:
#### Observation: Now the mean, min and max prices for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name are: 7.23, 2.4, 11.8. Since this result makes more sense, the price of product ID 21553, named 'Lowfat 2% Milkfat Cottage Cheese' will be replaced from 14.900 and inputed by the mean value mentioned above

In [ ]:
# Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese

df_orders_products_merged['prices'] = df_orders_products_merged['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [ ]:
# Check the statistics of the new 'prices' column in df_orders_products_merged

print('The statistics of the new prices column in df_orders_products_merged:')
df_orders_products_merged['prices'].describe()

In [ ]:
#### Observation: Now the mean, std, min and max results make more sense than Output [30]


# Step 5 (Updated: After this step --> replace({99999: 9.99, 14900: cottage_cheese['prices'].mean())

In [ ]:
#### Determine the difference between the spending habits of the three types of customers you identified.
#### Determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [ ]:
# Basic statistics of df_orders_products_merged by 'loyalty_flag' with the updated 'prices' column

print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_orders_products_merged:')
ords_prods_merge.groupby('Loyality Flag').agg({'prices': ['mean', 'min', 'max']})

In [ ]:
#### Observation: Although the result remains the same (compared to Output [29]): new customer holds the highest purchase price followed by regular customer then loyal customer, now the mean value becomes lower and max value is more reasonable (instead of 99999, now 25)

### Step 6

In [ ]:
#### Create a spending flag for each user based on the average price across all their orders using the following criteria:
    1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
    2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [ ]:
# Creating “average_price” column using transform()

ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [ ]:
ords_prods_merge.head()

In [ ]:
# Rounding averages 

ords_prods_merge['average_price'] = ords_prods_merge['average_price'].round(2)

In [ ]:
ords_prods_merge.head()

In [ ]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [ ]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [ ]:
ords_prods_merge.head()

In [ ]:
# Check the frequency
ords_prods_merge['spending_flag'].value_counts(dropna = False)

In [ ]:
### Step 7

In [ ]:
#### Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
    1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
    2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
    3. If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [ ]:
# Creating “median_prior_orders” column using transform()

ords_prods_merge['median_prior_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [ ]:
ords_prods_merge.head()

In [ ]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [ ]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[(ords_prods_merge['median_prior_orders'] > 10) & (ords_prods_merge['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [ ]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [ ]:
ords_prods_merge.head()

In [ ]:
# Check the frequency

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

In [ ]:
# Changing the null values in "frequency_flag" column to "Non-frequent customer"
ords_prods_merge['order_frequency_flag'].fillna("Non-frequent customer", inplace = True)

In [ ]:
ords_prods_merge['order_frequency_flag'].value_counts(dropna=False)

In [ ]:
# Check the null values

print('Records in df_orders_products_merged with null values in column order_frequency_flag:')
ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'median_prior_orders', 'order_frequency_flag']]

In [ ]:
# Check user_id 159838

ords_prods_merge[ords_prods_merge['user_id'] == 159838]

In [ ]:
print(ords_prods_merge[ords_prods_merge.order_frequency_flag.isnull()])

In [ ]:
#### Observation: user_id 159838 seems to be a new customer. Therefore, I leave the value of NaN as it is

In [ ]:
ords_prods_merge.shape

In [ ]:
### Step 8


In [ ]:
#### Ensure your notebook is clean and structured and that your code is well commented.

In [ ]:
#### Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.


In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))